### PR TITLE
Rp2040 countio

### PR DESCRIPTION
**Pull request by gamblor21**
Adding countio to RP2040. This build uses the PWM IRQ that does not seem in use in the PWM module. If the PWM module started to use it there could be a conflict.

### CODE REVIEW

File: ports/raspberrypi/common-hal/countio/Counter.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    pwm_config cfg = pwm_get_default_config();
+    pwm_config_set_clkdiv_mode(&cfg, PWM_DIV_B_RISING);
+    pwm_init(self->slice_num, &cfg, false);
+    gpio_set_function(self->pin_a, GPIO_FUNC_PWM);
+
+    self->count = 0;
+
+    claim_pin(pin_a);
+
+    MP_STATE_PORT(counting) = self;
```
Comment (tannewt) [inline ports/raspberrypi/common-hal/countio/Counter.c:36]: This won't work if you have multiple counters going at once. I think you want an array of size 8 (one for each slice) that can store each object. Then in the interrupt you'll need to check which slice wrapped and increment the corresponding counters count.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for this! The logic looks good except for storing the object. It won't work with multiple counters running at once. Suggestions inline.

File: ports/raspberrypi/common-hal/countio/Counter.c (hunk 1 above)
Comment (gamblor21) [reply ports/raspberrypi/common-hal/countio/Counter.c]: Fixed in the last commit

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the update! It looks good. However, I thought of another case that probably needs to be handled. We likely need to set one of the globals from PWMOut so that it doesn't allocate to the channel A pin at the same time. Could you try making a countio and then making a PWMOut on the other pin? It should error. Thanks!

**Review by tannewt — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**